### PR TITLE
[main] Port CHOMP to MoveIt2

### PR DESCRIPTION
**Pull request by andreas-botbuilt**
### Description

This PR makes [CHOMP](http://docs.ros.org/en/melodic/api/moveit_tutorials/html/index.html) available in MoveIt2. CHOMP can be used as a stand-alone motion planner plugin (instead of OMPL) or as a post-processor for OMPL (where the seed trajectory for CHOMP is the output of OMPL).

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

### Changes
- CHOMP parameters are loaded from the namespace "chomp"
- As suggested by @henningkayser in [PR 805](https://github.com/ros-planning/moveit2/pull/805), changes are made against `main`

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_planners/chomp/chomp_interface/package.xml
Comment (henningkayser) [inline moveit_planners/chomp/chomp_interface/package.xml:10]: I don't think Dave wants to be listed as maintainer here anymore. Please revert

File: moveit_planners/chomp/chomp_interface/CMakeLists.txt
Comment (henningkayser) [inline moveit_planners/chomp/chomp_interface/CMakeLists.txt:63]: I think I mentioned it in the original PR. It's fine to disable the tests for now, but please just comment them out and open an issue to get them fixed (doesn't have to be your job to fix them, though). All leftover TODOs should either reference an issue id or a GitHub username

File: moveit_planners/chomp/chomp_interface/CMakeLists.txt
Comment (henningkayser) [inline moveit_planners/chomp/chomp_interface/CMakeLists.txt:43]: @tylerjw didn't the export variable name change?

File: moveit_planners/chomp/chomp_interface/src/chomp_interface.cpp
Comment (henningkayser) [inline moveit_planners/chomp/chomp_interface/src/chomp_interface.cpp:72]: Since CHOMP is new in MoveIt 2, you can clean up things like this (just keep `collision_clearance`)

File: moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp
Comment (henningkayser) [inline moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp:45]: Can you make the LOGGER's `static const`? This should allow you to drop the numbers. Alternative to `static` you can also wrap the loggers in anonymous namespaces

File: moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp
Comment (henningkayser) [inline moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp:51]: Please use `std::make_shared` here like it was before

File: moveit_planners/chomp/chomp_interface/src/chomp_plugin.cpp
Comment (henningkayser) [inline moveit_planners/chomp/chomp_interface/src/chomp_plugin.cpp:71]: Same here

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:184]: same

File: moveit_planners/chomp/chomp_optimizer_adapter/src/chomp_optimizer_adapter.cpp
Comment (henningkayser) [inline moveit_planners/chomp/chomp_optimizer_adapter/src/chomp_optimizer_adapter.cpp:218]: This should better use `PLUGINLIB_EXPORT_CLASS`

**Review by henningkayser — COMMENTED** [review]
This looks really good so far, thanks a lot again for your work. I have left some minor fixes that should be addressed before merging, but otherwise I think this is good to go. Before merging: on what robot have you tested this (if anything else than the Panda)? I'm going to give this a couple runs and see if I find any issues.
I also noticed that there is still a bunch of commented code in this package (the parts not from you). I'm inclined to review these and simply remove commented code if it doesn't seem useful. Most of it seems to be related to profiling and debugging. Since you apparently use CHOMP currently for your project, feel free to re-enable and modify what you find or make suggestions for cleanup.

File: moveit_planners/chomp/chomp_interface/package.xml
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/package.xml:10]: Fixed.

File: moveit_planners/chomp/chomp_interface/src/chomp_interface.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/src/chomp_interface.cpp:72]: Great, fixed.

File: moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp:45]: Done.

File: moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp:51]: Fixed.

File: moveit_planners/chomp/chomp_interface/src/chomp_plugin.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/src/chomp_plugin.cpp:71]: Fixed.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:42]: These need a header guard, unfortunately (https://github.com/ros-planning/moveit2/runs/4263445115?check_suite_focus=true#step:8:361), since Rolling the geometry2 headers all switched to `.hpp`. See https://github.com/ros-planning/moveit2/blob/main/moveit_core/robot_state/src/robot_state.cpp#L43 for an example.

File: moveit_planners/chomp/chomp_optimizer_adapter/src/chomp_optimizer_adapter.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_optimizer_adapter/src/chomp_optimizer_adapter.cpp:218]: Fixed.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:184]: Fixed.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:49]: Fixed.

**Comment by andreas-botbuilt** [comment]
Thanks again for your review! It has been tested on a 6-DOF industrial robot arm and there have been no issues. I'm not using or planing to use the commented code at the moment so I agree that the profiling/debugging parts can be removed. However, it seems that the Hamiltonian Monte Carlo method has never been thoroughly tested (see [this comment](https://github.com/ros-planning/moveit2/blob/d03ae9b3c73b106566f86dcd4e8eafc6683c8e92/moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp#L171-L172)), so I would suggest to make that a github issue to be fixed in the future.

**Review by henningkayser — APPROVED** [review]
All right, to me this looks ready to get merged. @tylerjw or @JafarAbdi could either of you give this a quick review for the second approval?

**Comment by andreas-botbuilt** [comment]
Looks like I accidentally merged `main` into this, not sure if this was necessary.

File: moveit_planners/chomp/chomp_interface/package.xml
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_interface/package.xml:28]: This's no longer needed

File: moveit_planners/chomp/chomp_interface/package.xml
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_interface/package.xml:20]: I think this check is no longer needed, right ?

File: moveit_planners/chomp/chomp_interface/package.xml
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_interface/package.xml:9]: Why remove this line?

File: moveit_planners/chomp/chomp_interface/src/chomp_plugin.cpp
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_interface/src/chomp_plugin.cpp:67]: Why not use the input node ? also did you test having `~` as a namespace ? I think it will throw an exception

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:101]: No need to comment this, I think using clock from `std::chrono` should be sufficient 
```cpp
const auto wt = std::chrono::high_resolution_clock::now(); // Or system_clock
...
RCLCPP_INFO(LOGGER,"First coll check took " << std::chrono::duration<double>(std::chrono::high_resolution_clock::now() - wt).count());
```

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:304]: Same

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:435]: See my comment about using `chrono`

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:526]: See my comment about `chrono`

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:269]:
```suggestion
```

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:271]:
```suggestion
  res.processing_time_[0] = std::chrono::duration<double>(std::chrono::system_clock::now() - start_time).count();
```

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:296]:
```suggestion
  res.processing_time_.resize(1);
  res.processing_time_[0] = std::chrono::duration<double>(std::chrono::system_clock::now() - start_time).count();
```

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (JafarAbdi) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:294]: Any reason to add these lines ?

**Review by JafarAbdi — COMMENTED** [review]
Thanks for the port, just left a few comments then I think it's ready to be merged

**Review by JafarAbdi — COMMENTED** [review]
Just curious, there are a few reordering of include files and adding/removing black lines, any reason for that ?

File: moveit_planners/chomp/chomp_interface/CMakeLists.txt
Comment (JafarAbdi) [reply moveit_planners/chomp/chomp_interface/CMakeLists.txt:43]: We decided to use `${PROJECT_NAME}Targets`

File: moveit_planners/chomp/chomp_interface/package.xml
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/package.xml:28]: Fixed.

File: moveit_planners/chomp/chomp_interface/package.xml
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/package.xml:9]: Back in.

File: moveit_planners/chomp/chomp_interface/src/chomp_plugin.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/src/chomp_plugin.cpp:67]: This was closest to the original code. I removed the `~` namespace part, and it's now using the input node.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:101]: Fixed. I also put the runtime measurements at the end of `optimize()` back in (around line 523).

File: moveit_planners/chomp/chomp_interface/CMakeLists.txt
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/CMakeLists.txt:43]: Fixed.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:526]: Fixed.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:271]: Fixed.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:269]: Removed.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:296]: Fixed.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:294]: Fixed.

**Comment by andreas-botbuilt** [comment]
> Just curious, there are a few reordering of include files and adding/removing black lines, any reason for that ?

No reason. I'm not exactly sure what caused this, maybe some code formatting tool I used. If you prefer a specific ordering, let me know.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:432]: Looks like I skipped this one during my previous walk through, fixed now.

**Comment by JafarAbdi** [comment]
@andreas-botbuilt I open a [PR](https://github.com/andreas-botbuilt/moveit2/pull/1) to fix very nit stuff, could you please review it ? thanks

**Comment by JafarAbdi** [comment]
Sorry I forgot to run `clang-format`, do you mind fixing it? thanks!

**Review by JafarAbdi — APPROVED** [review]

**Review by henningkayser — APPROVED** [review]
lgtm

**Pull request merged**
